### PR TITLE
Refactor resnet

### PR DESCRIPTION
**Pull request by wenruifan**
### Description
- Refactor ResNet classes in image_cnn.py and add Flatten and Indentify modules.
- Add corresponding test code for new features.

#### Examples of new ResNet-related features:
1. Load ResNet and drop the last layer

Previous: 
```python
feature_network = torch.nn.Sequential(*(list(feature_network.children())[:-1]))
```

Current: 
```python
feature_network = torch.nn.Sequential(*(list(feature_network.model.children())[:-1]))
```

2. Generate ResNet series, e.g. ResNet18
Previous:
```python
class ResNet18Feature(nn.Module):
    def __init__(self, weights=models.ResNet18_Weights.DEFAULT):
        super(ResNet18Feature, self).__init__()
        model_resnet18 = models.resnet18(weights=weights)
        self.conv1 = model_resnet18.conv1
        self.bn1 = model_resnet18.bn1
        self.relu = model_resnet18.relu
        self.maxpool = model_resnet18.maxpool
        self.layer1 = model_resnet18.layer1
        self.layer2 = model_resnet18.layer2
        self.layer3 = model_resnet18.layer3
        self.layer4 = model_resnet18.layer4
        self.avgpool = model_resnet18.avgpool
        self._out_features = model_resnet18.fc.in_features

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)
        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)
        x = self.avgpool(x)
        x = x.view(x.size(0), -1)
        return x

    def output_size(self):
        return self._out_features
``` 
Current:
```python
class Flatten(nn.Module):
    def __init__(self):
        super(Flatten, self).__init__()

    def forward(self, x):
        return x.view(x.size(0), -1)

class ResNet18Feature(nn.Module):
    def __init__(self, weights=models.ResNet18_Weights.DEFAULT):
        super(ResNet18Feature, self).__init__()
        self.model = models.resnet18(weights=weights)
        self._out_features = self.model.fc.in_features
        self.model.fc = Flatten()

    def forward(self, x):
        return self.model(x)

    def output_size(self):
        return self._out_features
```

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: kale/embed/image_cnn.py [hunk 1]
```diff
@@ -8,6 +8,45 @@
...
+    def __init__(self):
+        super(Flatten, self).__init__()
+
+    def forward(self, x):
+        return x.view(x.size(0), -1)
+
+
+class Identity(nn.Module):
+    """Identity layer
+    This module is to any unwanted layer sof pre-trained models with
```
Comment (xianyuanliu) [inline kale/embed/image_cnn.py:33]: Maybe typo: sof

Comment (xianyuanliu) [reply kale/embed/image_cnn.py]: This module is to replace any unwanted layer of pre-trained models with

**Review by xianyuanliu — APPROVED** [review]
Nice simplification! Others need to review and align related work with this improvement.

File: kale/embed/image_cnn.py [hunk 2]
```diff
@@ -8,6 +8,44 @@
...
 
 
+class Flatten(nn.Module):
+    """Flatten layer
+    This module is to replace the last fc layer of the pre-trained model with a flatten layer. It flattens the input
+    tensor to 2D vector, which is (B, N). B is the batch size and N is the product of all dimensions except batch size.
+
+    Examples:
+        >>> x = torch.randn(8, 3, 224, 224)
+        >>> x = Flatten()(x)
```
Comment (haipinglu) [inline kale/embed/image_cnn.py:18]: Should this be `Flatten(x)` instead?

File: kale/embed/image_cnn.py [hunk 3]
```diff
@@ -8,6 +8,44 @@
 from torchvision import models
 
 
+class Flatten(nn.Module):
+    """Flatten layer
+    This module is to replace the last fc layer of the pre-trained model with a flatten layer. It flattens the input
+    tensor to 2D vector, which is (B, N). B is the batch size and N is the product of all dimensions except batch size.
```
Comment (haipinglu) [inline kale/embed/image_cnn.py:14]: Missing articles (English)
a 2D vector
except the batch size

File: kale/embed/image_cnn.py [hunk 4]
```diff
@@ -8,6 +8,44 @@
...
+    def __init__(self):
+        super(Flatten, self).__init__()
+
+    def forward(self, x):
+        return x.view(x.size(0), -1)
+
+
+class Identity(nn.Module):
+    """Identity layer
+    This module is to replace an unwanted layers of the pre-trained model with an identity layer.
```
Comment (haipinglu) [inline kale/embed/image_cnn.py:32]: English: an (singular)  .... layers (plural)?

File: kale/embed/image_cnn.py [hunk 5]
```diff
@@ -8,6 +8,44 @@
...
+    It returns the input tensor as output.
```
Comment (haipinglu) [inline kale/embed/image_cnn.py:33]: the output

**Review by haipinglu — CHANGES_REQUESTED** [review]
Some minor English problems to fix before merging. 
Also, clarify the ToDos in the description please.

**Review by haipinglu — APPROVED** [review]
Well done!

**Pull request merged**